# Проект по SQL

## Описание проекта

### Контекст

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

###  Исследование

#### Задачи исследования

- Посчитайть, сколько книг вышло после 1 января 2000 года;

- Для каждой книги посчитайть количество обзоров и среднюю оценку;

- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

- Определить автора с самой высокой средней оценкой книг — учитывайть только книги с 50 и более оценками;

- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

#### Ход исследования

- Описать цели исследования;
- Исследовать таблицы — вывести первые строки;
- Сделайть по одному SQL-запросу для решения каждого задания;
- Вывести результаты каждого запроса в тетрадке;
- Описать выводы по каждой из решённых задач.

### Обзор данных

#### Описание данных

Таблица books

Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.


Таблица authors

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.


Таблица publishers

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица ratings

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.


Таблица reviews

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

#### Схема данных

<img src=https://pictures.s3.yandex.net/resources/scheme_1589269096.png><br>

## Загрузки данных и изучение общей информации

### Загрузка данных

#### Подключим библиотеки необходимые для работы

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

**Устанавливаем параметры**

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

**Сохраняем коннектор**

In [3]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**Установим опции по выводу в датафрейме**

In [4]:
pd.set_option('display.float_format', '{:.2f}'.format) #все числа отражаются с двумя знаками после запятой
pd.set_option('display.max_colwidth', None) # для полноформатного отборажения таблиц
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

#### Откроем таблицы

In [5]:
list= ['books', 'authors', 'publishers', 'ratings', 'reviews']
for i in list:
    df = pd.io.sql.read_sql('''SELECT * FROM {}'''.format(i), con = engine)
    print(i,'\n')
    display(df.head())
    print('\n')

books 



book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268



authors 



author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd



publishers 



publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company



ratings 



rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2



reviews 



review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.

#### Общая информация по датафрейму

In [6]:
list= ['books', 'authors', 'publishers', 'ratings', 'reviews']
for i in list:
    df = pd.io.sql.read_sql('''SELECT * FROM {}'''.format(i), con = engine)
    print(i,'\n')
    print(df.info())
    print('\n')

books 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None


authors 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None


publishers 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column     

**Вывод**

Все открылись. Таблицы без пропусков, типы данных менять не надо. 

## Исследовательский анализ данных

### Cколько книг вышло после 1 января 2000 года

In [7]:
query = '''
            SELECT COUNT(book_id)
            FROM books
            WHERE publication_date > '2000-01-01'
        '''
con=engine.connect()
df = pd.io.sql.read_sql(sql=text(query), con = con)
df

count
0    819

**Вывод**

После 1 января 2000 года вышло 819 книг.

### Количество обзоров и средняя оценка для каждой книги

In [8]:
query = '''
            SELECT b.title,
                    COUNT(DISTINCT e.review_id) AS reviews,
                    ROUND(AVG(a.rating), 2) AS avg_rating                   
            FROM books b
            LEFT JOIN reviews e ON b.book_id = e.book_id
            LEFT JOIN ratings a ON b.book_id = a.book_id
            GROUP BY b.title
            ORDER BY reviews DESC
        '''
con=engine.connect()
df = pd.io.sql.read_sql(sql=text(query), con = con)
df          

title  \
0                                                                                                                                                    Memoirs of a Geisha   
1                                                                                                                                                Twilight (Twilight  #1)   
2                                                                                                             Harry Potter and the Chamber of Secrets (Harry Potter  #2)   
3                                                                                                                                                       The Glass Castle   
4                                                                                                                                                        Eat  Pray  Love   
5                                                                                                                      The Curious Incident of the Dog in the Night-Time   
6                                                                                                                                                    Water for Elephants   
7                                                                                                                                              Outlander (Outlander  #1)   
8                                                                                                                                                               The Road   
9                                                                                                                                                          The Alchemist   
10                                                                                                                                The Da Vinci Code (Robert Langdon  #2)   
11                                                                                                                                                The Catcher in the Rye   
12                                                                                                                                   The Hobbit  or There and Back Again   
13                                                                                                             The Lightning Thief (Percy Jackson and the Olympians  #1)   
14                                                                                                           Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)   
15                                                                                                                                             The Giver (The Giver  #1)   
16                                                                                                                                                        The Book Thief   
17                                                                                                                                                    My Sister's Keeper   
18                                                                                                                                                              Coraline   
19                                                                                                                                                     Lord of the Flies   
20                                                                                                                                                     Holes (Holes  #1)   
21                                                                                                                                              A Tree Grows in Brooklyn   
22                                                                                                                                                               Dracula   
23                                 

**Вывод**

Максимально количество обзоров набрала книга Memoirs of a Geisha - 8 шт, со средним рейтингом 4,14.

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

Необходимо учитывайть только книги толще 50 страниц — так мы исключим из анализа брошюры.

In [9]:
query = '''
            SELECT p.publisher,
                    COUNT(b.book_id) AS count_books
            FROM publishers p
            LEFT JOIN books b ON p.publisher_id = b.publisher_id
            WHERE b.num_pages > 50
            GROUP BY p.publisher
            ORDER BY count_books DESC
            LIMIT 1
        '''

con=engine.connect()
df = pd.io.sql.read_sql(sql=text(query), con = con)
df 

publisher  count_books
0  Penguin Books           42

**Вывод**

Издательство, которое выпустило наибольшее число книг - Penguin Books.

### Автор с самой высокой средней оценкой книг

Необходимо учитывайть только книги с 50 и более оценками.

In [10]:
query = '''
            SELECT a.author,
                    AVG(r.rating) AS avg_rating
            FROM authors a
            LEFT JOIN books b ON a.author_id = b.author_id
            LEFT JOIN ratings r ON b.book_id = r.book_id
            WHERE b.book_id IN (SELECT book_id
                                FROM ratings
                                GROUP BY book_id
                                HAVING COUNT(rating) >=50)
            GROUP BY a.author   
            ORDER BY avg_rating DESC
            LIMIT 1
                                
        '''

con=engine.connect()
df = pd.io.sql.read_sql(sql=text(query), con = con)
df 

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

**Вывод**

Автор J.K. Rowling/Mary GrandPré имеет самую высокую среднюю оценку - 4,29.

### Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [11]:
query = '''
                SELECT FLOOR(AVG(g.reviews)) AS avg_reviews
                FROM (
                        SELECT username,
                                COUNT(DISTINCT review_id) AS reviews
                        FROM reviews
                        WHERE username IN (
                                        SELECT username
                                        FROM ratings
                                        GROUP BY username
                                        HAVING COUNT(rating) > 48
                                        ) 
                        GROUP BY username
                ) AS g
        '''

con=engine.connect()
df = pd.io.sql.read_sql(sql=text(query), con = con)
df 

avg_reviews
0        24.00

**Вывод**

Среднее количество обзоров от пользователей - 24.

## Выводы

Открыли все датафреймы - пропусков нет и типы данных менять не надо.

Итоги по исследованиям:

- После 1 января 2000 года вышло 819 книга;
- Максимально количество обзоров набрала книга Memoirs of a Geisha - 8 шт, со средним рейтингом 4,14;
- Издательство, которое выпустило наибольшее число книг - Penguin Books;
- Автор J.K. Rowling/Mary GrandPré имеет самую высокую среднюю оценку - 4,29;
- Среднее количество обзоров от пользователей - 24.